File: 1718-construct-the-lexicographically-largest-valid-sequence/1718-construct-the-lexicographically-largest-valid-sequence.cs
public class Solution {
    int len;
    int[] ans;
    bool[] visited;

    public int[] ConstructDistancedSequence(int n) {
        if(n == 1) return [1];

        len = 2 * (n-1) + 1;
        ans = new int[len];
        visited = new bool[n+1];

        PlaceNums(n, 0);
        return ans;
    }

    // place the best possible, biggest unplaced num at "idx"
    private bool PlaceNums(int n, int idx){
        if(idx == len) return true;     // filled all nums

        for(var num=n; num>0; num--){
            if(visited[num]) continue;  // this num is already filled in ans

            var secIdx = num == 1 ? idx : idx+num;  // second index of curr num
            if(secIdx >= len || ans[idx] != 0 || ans[secIdx] != 0) continue;

            // place current num
            ans[idx] = ans[secIdx] = num;
            visited[num] = true;

            // place remaining nums
            int nextIdx = 0;
            while(nextIdx < len && ans[nextIdx] != 0) nextIdx++;
            if(PlaceNums(n, nextIdx)) return true;

            // rollback current num placement
            visited[num] = false;
            ans[idx] = ans[secIdx] = 0;
        }
        return false;
    }
}

File: 1718-construct-the-lexicographically-largest-valid-sequence/1718-construct-the-lexicographically-largest-valid-sequence.py
class Solution:
    def constructDistancedSequence(self, n: int) -> List[int]:
        if n == 1: return [1]

        self.lenn = 2 * (n-1) + 1
        self.ans = [0] * self.lenn
        self.visited = [False] * (n+1)

        self.placeNums(n, 0)
        return self.ans

    # place the best possible, biggest unplaced num at "idx"
    def placeNums(self, n, idx):
        if idx == self.lenn: return True     # filled all nums

        for num in range(n, 0, -1):
            if self.visited[num]: continue  # this num is already filled in ans

            secIdx = idx if num == 1 else idx+num  # second index of curr num
            if secIdx >= self.lenn or self.ans[idx] != 0 or self.ans[secIdx] != 0: continue

            # place current num
            self.ans[idx] = self.ans[secIdx] = num
            self.visited[num] = True

            # place remaining nums
            nextIdx = 0
            while nextIdx < self.lenn and self.ans[nextIdx] != 0: nextIdx += 1
            if self.placeNums(n, nextIdx): return True

            # rollback current num placement
            self.visited[num] = False
            self.ans[idx] = self.ans[secIdx] = 0
        return False